matchmaking/matchmaker.py

```diff
@@ -0,0 +1,344 @@
+
+        self._recent_match_by_profile_pair = dict()
+        self._recent_match_by_profile_topic = dict()
+        self._recent_match_by_profile_pair_and_topic = dict()
+        self._match_requests_profile_to_topic = dict()
+        self._match_requests_topic_to_profile = dict()
+        self._possible_matches = dict()
+
+    def _gather_data(self):
+        "pull the data from the database and create lookup dicts used in scoring and filtering to admissible matches"
+        match_requests = MatchRequest.objects.filter(date__gte=self.match_request_since_date)\
+            .values_list('profile__email', 'topic_channel__name')
+        emails = set(mr[0] for mr in match_requests)
+
+        # recent matches
+        matches = Match.objects.filter(date__gte=self.match_since_date)\
+            .filter(profiles__email__in=emails)\
+            .order_by('date')\
+            .prefetch_related('profiles', 'topic_channel')
+        most_recent_match_by_profile = {}
+        recent_match_by_profile_pair = {}
+        recent_match_by_profile_topic = {}
+        recent_match_by_profile_pair_and_topic = {}
+        for match in matches:
+            profiles = list(match.profiles.all().values_list('email', flat=True))
+            data = {'num_attending': len(profiles), 'date': match.date}
+            topic = match.topic_channel.name
+            for i in range(len(profiles)):
+                prof_key = key(profiles[i])
+                if (
+                    prof_key not in most_recent_match_by_profile or
+                    most_recent_match_by_profile[prof_key]['date'] > data['date']
+                ):
+                    most_recent_match_by_profile[prof_key] = data
+                recent_match_by_profile_topic[key(profiles[i], topic)] = data
+                for j in range(i + 1, len(profiles)):
+                    recent_match_by_profile_pair[key(profiles[i], profiles[j])] = data
+                    recent_match_by_profile_pair_and_topic[key(profiles[i], profiles[j], topic)] = data
+
+        # match requests lookups
+        match_requests_profile_to_topic = {}
+        match_requests_topic_to_profile = {}
+        for match_request in match_requests:
+            match_requests_profile_to_topic.setdefault(match_request[0], set()).add(match_request[1])
+            match_requests_topic_to_profile.setdefault(match_request[1], set()).add(match_request[0])
+
+        possible_matches = {}
+        for profile_A, topics in match_requests_profile_to_topic.items():
+            possible_matches[profile_A] = set()
+            for topic in topics:
+                profiles_B = match_requests_topic_to_profile[topic]
+                for profile_B in profiles_B:
+                    if profile_A != profile_B:
+                        possible_matches[profile_A].add(profile_B)
+
+        ######################
+        # assign instance vars
+        self._most_recent_match_by_profile = most_recent_match_by_profile
```

Comment:
Each method in this class is designed to be responsible for as little as possible, so that it's easier to reason about. I _could_ start processing data in `_gather_data`, but then it wouldn't be `_gather_data` it would be `_gather_data_and_start_figuring_out_topic_stuff`... but then it would be hard for a future programmer to know where to best find the code responsible for whatever they are looking at, and it become hard to know where to change the code. ... This method just gathers data.